packages/venia-ui/lib/components/ProductFullDetail/customAttributes.js

```diff
@@ -0,0 +1,251 @@
+import React, { useMemo } from 'react';
+import { array, shape, string } from 'prop-types';
+import { FormattedDate, FormattedMessage, FormattedTime } from 'react-intl';
+
+import { BrowserPersistence } from '@magento/peregrine/lib/util';
+import { useStyle } from '@magento/venia-ui/lib/classify';
+import Price from '@magento/venia-ui/lib/components/Price';
+import RichContent from '@magento/venia-ui/lib/components/RichContent';
+
+import defaultClasses from './customAttributes.module.css';
+
+const storage = new BrowserPersistence();
+
+const CustomAttributes = props => {
+    const { customAttributes } = props;
+    const classes = useStyle(defaultClasses, props.classes);
+
+    // TODO: Get complete price data from GraphQl including currency
+    const storeCurrency = storage.getItem('store_view_currency') || 'USD';
+
+    // Remove attributes not visible on front
+    const cleanAttributes = useMemo(() => {
+        if (customAttributes.length === 0) {
+            return [];
+        }
+
+        return customAttributes
+            .filter(
+                ({ attribute_metadata }) =>
+                    attribute_metadata.is_visible_on_front === true
+            )
+            .sort(
+                (a, b) =>
+                    a.attribute_metadata.sort_order -
+                    b.attribute_metadata.sort_order
+            );
+    }, [customAttributes]);
+
+    const list = useMemo(() => {
+        if (cleanAttributes.length === 0) {
+            return [];
+        }
+
+        return cleanAttributes.map(
+            ({
+                selected_attribute_options,
+                entered_attribute_value,
+                attribute_metadata
+            }) => {
+                const key = attribute_metadata.uid;
+                const attributeLabel = attribute_metadata.label;
+                const {
+                    ui_input_type: type,
+                    is_html_allowed: isHtml
+                } = attribute_metadata.ui_input;
+                let attributeContent;
+
+                switch (type) {
```

Comment:
I would suggest extracting this logic into its own component, and let a supporting talon handle as much as possible. 

Optional, but definitely valuable for extensibility, I would maybe even take this a step further and apply factory pattern here (`<AttributeFactory attribute={attribute} />`), and break each of these cases up into tiny components.